text/0000-ssr-stylesheet-deduplication.md

```diff
@@ -0,0 +1,165 @@
+
+### API shape
+
+A new property bag is added to the `renderComponent` API of `@lwc/ssr-runtime`.
+
+```ts
+function renderComponent(
+    tagName: string, 
+    Constructor: LightningElement, 
+    props: any,
+    // The below is new
+    options?: {
+        stylesheetCache?: Set<String>;
+    }
+): void;
+```
+
+For the client side, consumers are expected to import the `registerLwcStyleComponent` function and call it once:
+
+```js
+import { registerLwcStyleComponent } from '@lwc/ssr-runtime'
+registerLwcStyleComponent()
+```
+
+Under the hood, this will call `customElements.define()` with a vanilla `<lwc-style>` component whose job is to identify nearby `<style>`s and push them into either `shadowRoot.adoptedStyleSheets` or `document.adoptedStyleSheets` (for top-level light DOM).
+
+Consumers should most likely put this `<script>` into the `<head>` so that it registers the web component as soon as possible.
+
+### Prior art
+
+#### Web component frameworks
+
+Lit has [toyed with the idea](https://github.com/lit/lit/issues/4357) of deduplicating styles, but not committed to anything. The two high-level ideas they have explored are the `<link rel=stylesheet>` solution and the "inline web component" solution.
+
+FAST has implemented a ["styled component"](https://github.com/microsoft/fast/pull/5600) approach relying on an inline web component to copy styles from a global source into individual components. (Notably this approach has led to [Flash of Unstyled Content issues](https://github.com/microsoft/fast/issues/5906), but this seems to be an implementation issue and not a problem inherent in the design.)
+
+The Microsoft Edge team has [an explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/ShadowDOM/explainer.md) proposing a solution to this problem at the level of HTML, but this is still far from being available cross-browser. And when it is, it is unclear how backwards compatibility with older browsers should work.
+
+#### Non-web-component frameworks
+
+Most non-web-component using frameworks (such as Vue, Svelte, and React) have a system where styles can be extracted from compiled components and are assumed to be hoisted to the document `<head>`. In fact, in Vite, that is what this does:
+
+```js
+import './stylesheet.css' // auto-hoists
+```
+
+Needless to say, this is insufficient for our use case since LWC has first-class support for shadow DOM components.
+
+## Drawbacks
+
+This solution will add significant complexity both to the LWC engine (which will have to deal with supporting the new API in SSR, as well as potentially deduplicating styles client-side w.r.t the current CSR implementation) and on the consumer side (which will have to do the bulk of the work of managing and hosting the stylesheets themselves).

```

Comment:
> consumer side (which will have to do the bulk of the work of managing and hosting the stylesheets themselves)

Does the consumer need to add `<style>` or `<link>` tags to the base doc for each `stylesheetCache` item? Or does the `<lwc-style>` component handle the stylesheets (as long as `registerLwcStyleComponent()` is called on the client)?

Reply:
No, the consumer does not need to do anything. LWC itself (and notably the `<lwc-style>` component) will 1) render a `<style>` the first time it sees it, and put it in a global cache, and then 2) render nothing the next time, because it can look up the `<style>` via the `style-id` attribute. I'd recommend reading [Rob Eisenberg's design](https://eisenbergeffect.medium.com/sharing-styles-in-declarative-shadow-dom-c5bf84ffd311) which I shamelessly stole.

Reply:
>on the consumer side (which will have to do the bulk of the work of managing and hosting the stylesheets themselves).

I am confused about this statement relative to the above comments.

Reply:
+1. That statement triggered my original question in this thread.

Reply:
Invalid sentence from the previous RFC. Removed.